### DISCUSSION TITLE
[May 5 Discussion] A Type and Effect System for Deterministic Parallel Java

### DISCUSSION THREAD

**Discussion by susan-garry** [sampsyo/cs6120#324] [Reading]
Here is the discussion thread for [A Type and Effect System for Deterministic Parallel Java](http://dpj.cs.illinois.edu/DPJ/Publications_files/DPJ-OOPSLA-2009.pdf).
@barabanshek (Nikita Lazarev) and I will be leading this discussion. Please post your thoughts and questions below!

**Comment by 5hubh4m** [comment]
My question about this paper is a classic one: what are you willing to pay for formal guarantees?

This paper wants to solve the problem of writing fully generic parallel code with a deterministic guarantees, which is a really non-trivial problem to solve, so I wasn't surprised to see that their solution is kind of alien-looking, from the perspective of a normal programmer. Manually segregating heap into various regions lets them prevent nasty concurrency bugs but it increases the mental burden on the programmer.

There are definitely better abstractions that make writing parallel code easy, even though they might not necessarily offer formal guarantees. Curious to hear what others think about the usability of their language.

**Reply by barabanshek** [reply]
Great comment! Generally any (almost) abstraction/language feature that is meant to prevent bugs introduces certain mental burden. Think of smart pointers in C++: they do prevent memory leak bugs, but programmers need to understand them well enough and use properly; in fact, there are at least three types of smart pointers with their own limitations and use cases. Just google 'smart pointers c++ tutorial' and you will see 256000 results. But once you get them, you C++ code indeed becomes much less error prone in terms of memory management and this has the benefits in the long run. But everything has the limit, and at some point these language extensions become just too much or not really appropriate for a given language. I think, this is what's happening in this paper.

If one really needs to make sure their code is thread-safe, then paradigms with immutable variables like functional languages and similar work really best here.  For example, I'm sure their array abstraction (which is pretty limited btw) can be perfectly replaced by map-reduce for most of the applications they had in mind. But Java is an imperative language by concept. Imperative languages are not the best to express concurrency, so any attempts to make them deterministically parallel (not even talking about being formally verified) will be complicated (imho). As a programmer I would've chosen, say, Scala for implementing a highly-concurrent algorithm, and this way it will also be way easier to formally prove that it is indeed safe if necessary. I don't really see a point of making Java deterministically parallel in general, not only in the way they do. And the complications (and limitations) of their approach is a good proof of that.

**Reply by 5hubh4m** [reply]
I was also thinking of Scala/actors when I was talking about abstractions for writing asynchronous code.

**Reply by 5hubh4m** [reply]
I agree with functional programming being a great way to write code that can be transparently parallelized!

**Reply by sampsyo** [reply]
I think it's a good question to ask about the cost/benefit analysis of determinism. It would be worth weighing one against the other—but it's hard because the programming difficulty is sort of a short-term concern while latent concurrency heisenbugs are a long-term one.

I'm not sure I understand the distinction y'all are drawing between Scala and Java. Actors are cool, but you can do actors in other languages too. And they do not address determinism. You might get "safety" by other definitions, but are you thinking of a specific way in which they help with determinism in particular?

**Reply by 5hubh4m** [reply]
Not talking about Scala per se but actors which I think Scala helped popularize as a concept. Actors don’t give you determinism but make writing concurrent code easier since it feels like a “natural” extension of OOP anyway.

**Reply by anshumanmohan** [reply]
This is clearly an important domain with some decent-ish solutions. Among other things, this work exposes the laundry list of things one needs to think about in order to [actually get it right](https://youtu.be/k6hZ9KdG1QU). I think there's value to that. I agree that the syntax presented is too clunky and low-level for regular use, but perhaps DPJ can be used as an IR or an oracle when compiling a cleaner higher-level language?

**Reply by ayakayorihiro** [reply]
I agree with the general points made in this thread, namely that this is definitely a good step towards solving the very difficult and important problem but the tradeoff burden on the programmer is a bit heavier that one would (potentially) hope. I do think the amount of annotations that one must put in for this to work is pretty large, considering that in a case in these benchmarking programs (which by themselves aren't too large in terms of SLOC), you'd have to annotate ~10.7% of your code (22.6% in the worst case). I think something that would be very interesting to see is a potential user study on DPJ, to obtain some metric on how much burden would be on a programmer that wanted to have these formal guarantees. 

Also @anshumanmohan nice parks reference :smile:

**Reply by charles-rs** [reply]
I was also thinking of what it would look like to study the effects of something like DPJ! I think one of the best settings would be some kind of corporate environment, where there is a focus on "getting things done" and to record how programmers spend their time (debugging bad parallel code vs annotating their parallel code to make it deterministic).

I think that the determinism aspect is worth paying quite a price for in a large system, as it has the advantage of meaning that your testing is a LOT more representative of the real world, especially when there are issues across machines (i.e. it was tested on a fast computer, so thread A ALWAYS terminates before thread B has a chance to get to a certain point, but not so on a slow proc).. 

In contrast to some opinions above, I don't think the syntax is really that bad; Java is pretty verbose already, so it doesn't seem like it's making it much worse.

**Reply by alaiasolkobreslin** [reply]
I was also thinking about the usability of a language that would require so much code to contain annotations. I bet the amount of time it would take programmers to be able to reason about all of these annotations would be very high, and it would have been interesting to see a user study to confirm this.

The authors briefly mention a tool in development called DPJIZER that infers many annotations using iterative constraint solving. In this [paper](https://dpj.cs.illinois.edu/DPJ/Publications_files/DPJizer-ASE-2009.pdf), they describe a user study that showed across 11 programs, DPJIZER saved the writing of 406 effects manually. They conclude that DPJIZER drastically reduces the number of annotations that need to be written manually, and the automatically inferred effects are more precise. So it's cool that they have tried to tackle this issue of usability.

**Comment by orkosinha** [comment]
I thought this paper was interesting, as I've never thought of this kind of approach to solving concurrency problems. This reminded me of [Harmony](https://harmony.cs.cornell.edu/) which aims to solve the issue of "The parallel programmer today prunes away the nondeterminism using constructs such as locks and semaphores, then debugs the program to eliminate the symptoms. This task is. tedious, error prone, and extremely challenging even with good debugging tools." My question is at what point is using something like Harmony, which checks all interleaving of instructions, better or worse to use than something like DPJ

**Reply by tonyjie** [reply]
I think Harmony is a model checking tool for nondeterminism debugging, while DPJ is really working on OS-level? For example, an `atomic` instruction in Harmony would not lead to a real `atomic` load/store/xxx instruction running on OS, but just a kind of simulation, while DPJ and corresponding type and effect system is a real system that interacts with OS. I think that's the main difference. 

Please correct me if there's anything wrong!

**Reply by chhzh123** [reply]
I for the first time know Harmony has this cool feature. I think both Harmony and DPJ share the same goal to eliminate data racing and undefined behavior. From my point of view, it seems the checking in Harmony is more like a single brute-force pass to enumerate all the possible instruction interleaving, while DPJ enforces the determinism in its type and effect system but requires programmers to provide more information when writing concurrent programs.

**Reply by JonathanDLTran** [reply]
I imagine that the Harmony model checking tool checks every possible instruction ordering, and likely scales poorly the larger the program size gets. On the other hand, DPJ may reject programs that do not contain data races, but which the compiler cannot guarantee via the given type system. But DPJ can accept larger programs as well, as long as the types the programmer gives allow the type system to show that determinism can be achieved. I think it would be interesting to see if the model checking idea can be combined with the type -effect system proposed, to allow even more programs to be compiled.

**Comment by andreyyao** [comment]
Here are my first impressions after reading the paper. First of all, this partial solution to the nondeterministic parallelism problem follows the philosophy of "programmers helping the compiler by telling it statically stuff it wouldn't be able to figure out itself", which I think is a necessary approach when your target program behavior specification becomes too complicated. Secondly, this paper feels somewhat similar to the Typed Based Alias Analysis paper.

The constructs they defined to express nested regions(with `:`), arbitrarily deep descendants of regions(with `:`), and others are mostly intuitive. I wonder if it's possible to unify the tree-like structures for nested regions and the subarray stuff(where you can have different partitions of the same array) by describing the set of regions as a bounded lattice or something

**Comment by gsvic** [comment]
The authors present a method for expressing deterministic parallel algorithms in imperative programming languages. Providing deterministic guarantees for such languages is challenging due to the complexity of the various features they provide, like aliasing and references, and proving that the synchronization mechanism preserves determinism can be tricky. The main idea is based on a type and effect system which segments the heap into regions, allowing the developer to define the access types (reads/writes) to these regions, enabling the compiler to check if the synchronization ensures determinism. They also provide a Java extension, namely DPJ (Deterministic Parallel Java), that applies the ideas presented in the paper in Java. 

I found a lot of technical details in the paper interesting, like the hierarchical partitioning of the heap using RPLs. While DPJ does not look so complex in the examples presented in the paper (like the quicksort example), I am wondering about the ease of use in more complex use-cases.

**Comment by atucker** [comment]
I might be misunderstanding the paper, but I'm confused by why they emphasize the deterministic nature of "most" programs in an input-output sense, over the deterministic nature of "most" programs in terms of memory accesses. If two threads are in a race condition, then it doesn't matter that the threads computing a deterministic function of their inputs, it matters that they might access the memory in different orders. Their solution of separating the heap into different regions and then making sure that they don't overlap is definitely addressing the memory access problem.

**Reply by andrewb1999** [reply]
I was confused about this at first too, but I think what they're really trying to describe is "deterministic parallelism", rather than deterministic parallel java actually being deterministic on "most" or all programs. There are of course algorithms that are non-deterministic by default (simple example, calling a random number generator), and I believe these algorithms would be allowed in DPJ. I believe the claim they're making is that deterministic algorithms that work under DPJ will be guaranteed to be deterministic.

These are the same claims that other languages with compile time guarantees about concurrency, like rust, try to make.
